Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift

```diff
@@ -0,0 +1,122 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension AsyncSequence where Self: Sendable {
+  /// Creates an asynchronous sequence that buffers elements.
+  ///
+  /// The buffering behaviour is dictated by the policy:
+  /// - bounded: will buffer elements until the limit is reached. Then it will suspend the upstream async sequence.
+  /// - unbounded: will buffer elements without limit.
+  /// - bufferingLatest: will buffer elements until the limit is reached. Then it will discard the oldest elements.
+  /// - bufferingOldest: will buffer elements until the limit is reached. Then it will discard the newest elements.
+  ///
+  /// - Parameter policy: A policy that drives the behaviour of the ``AsyncBufferSequence``
+  /// - Returns: An asynchronous sequence that buffers elements up to a given limit.
+  public func buffer(
+    policy: AsyncBufferSequence<Self>.Policy
```

Comment:
I am wondering if this is the right way to do this. Didn't we last discuss to move the `Policy` into its standalone top level type called `AsyncBufferSequencePolicy`?. In the case where you just get a sequence and want to call `.buffer(policy: .unlimited)` this works great, but if you want to spell hold the policy somewhere. It is getting hard to spell it out. I am inclined to say we should love it to its own top level type.

Reply:
I wonder if the policy system is perhaps something we need for other things too. I agree that a buffering type being coupled with its policy seems appropriate (specifically the way it is currently written). But, if the policy system is general to any buffering mechanism then it should be at the top level.